src/go/util/httppattern/match_sequence_generator.go

```diff
@@ -0,0 +1,194 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httppattern
+
+import (
+	"fmt"
+	"strings"
+)
+
+// MatchSequenceGenerator generate the match sequence of added methods based on their
+// http pattern. The sequence should be same as the matching order of
+// https://github.com/GoogleCloudPlatform/esp-v2/blob/641ce1d5c177401e424f2b27dd45de1bf797530b/src/api_proxy/path_matcher/path_matcher.h#L1
+type MatchSequenceGenerator struct {
+	RootPtr     *matchSequenceGeneratorNode
+	CustomVerbs map[string]bool
+}
+
+type Method struct {
+	UriTemplate string
+	HttpMethod  string
+	Operation   string
+}
+
+func NewMatchSequenceGenerator() *MatchSequenceGenerator {
+	return &MatchSequenceGenerator{
+		RootPtr:     newMatchSequenceGeneratorNode(),
+		CustomVerbs: make(map[string]bool),
+	}
+}
+
+func (h *MatchSequenceGenerator) Register(method *Method) error {
+	if method == nil {
+		return fmt.Errorf("empty method")
+	}
+
+	uriTemplate := method.UriTemplate
+	httpMethod := method.HttpMethod
+
+	ht := Parse(uriTemplate)
+	if ht == nil {
+		return fmt.Errorf("invalid url template `%s`", uriTemplate)
+	}
+
+	pathInfo := transferFromUriTemplate(ht)
+	methodData := &methodData{
+		Method:   method,
+		Variable: ht.Variables,
+	}
+
+	if !h.RootPtr.insertPath(pathInfo, httpMethod, methodData, true) {
+		return fmt.Errorf("duplicate http pattern `%s %s`", httpMethod, uriTemplate)
+	}
+
+	if ht.Verb != "" {
+		h.CustomVerbs[ht.Verb] = true
+	}
+
+	return nil
+}
+
+type MatchSequence []*Method
+
+// Return a sorted slice of methods, used to match incoming request in sequence.
+func (h *MatchSequenceGenerator) Generate() *MatchSequence {
+	result := &MatchSequence{}
+	h.RootPtr.traverse(result)
+	return result
+
+}
+
+type variableBinding struct {
+	FieldPath []string
+	Value     string
+}
+
+type methodData struct {
+	*Method
+	Variable []*variable
+}
+
+type lookupResult struct {
+	data       *methodData
+	isMultiple bool
+}
+
+func (sr *MatchSequence) appendMethod(m *Method) {
+	if m != nil {
+		*sr = append(*sr, m)
+	}
+}
+
+/////////////////////////////////////////////
+// following lookup code are for test-only //
+/////////////////////////////////////////////
+func (h *MatchSequenceGenerator) lookup(httpMethod string, path string) (*Method, []*variableBinding) {
+	parts := extractRequestParts(path, h.CustomVerbs)
+
+	if h.RootPtr == nil {
+		return nil, nil
+	}
+
+	result := &lookupResult{}
+	h.RootPtr.lookupPath(parts, 0, httpMethod, result)
+	if result.data == nil {
+		return nil, nil
+	}
+
+	return result.data.Method, extractBindingsFromPath(result.data.Variable, parts)
+}
+
+func extractRequestParts(path string, customVerbs map[string]bool) []string {
+	// Remove query parameters.
+	if idx := strings.Index(path, "?"); idx != -1 {
```

Comment:
Can we make use of the http_template parser instead of doing this manual string manipulation?

Or i guess I'm confused about the order. When will the http template be parsed, when will this generator be used?

Reply:
Done.